hudi-common/src/main/java/org/apache/hudi/common/util/HoodieRecordUtils.java

```diff
@@ -75,21 +85,58 @@ public static HoodieRecordMerger loadRecordMerger(String mergerClass) {
    * Instantiate a given class with a record merge.
    */
   public static HoodieRecordMerger createRecordMerger(String basePath, EngineType engineType,
-                                                      List<String> mergerClassList, String recordMergerStrategy) {
-    if (mergerClassList.isEmpty() || HoodieTableMetadata.isMetadataTable(basePath)) {
+                                                      List<String> mergerClassList,
+                                                      String recordMergerStrategy,
+                                                      Option<String> payloadClassOpt) {
+    if (HoodieTableMetadata.isMetadataTable(basePath)) {
       return HoodieAvroRecordMerger.INSTANCE;
+    } else if (mergerClassList.isEmpty()) {
+      // No merger class is given, we fall back to use Avro based mergers.
+      return createValidRecordMerger(engineType, "", recordMergerStrategy, payloadClassOpt)
+          .orElse(HoodieAvroRecordMerger.INSTANCE);
     } else {
       return createValidRecordMerger(engineType, mergerClassList, recordMergerStrategy)
           .orElse(HoodieAvroRecordMerger.INSTANCE);
     }
   }
 
   public static Option<HoodieRecordMerger> createValidRecordMerger(EngineType engineType,
-                                                                   String mergerImpls, String recordMergerStrategy) {
-    if (recordMergerStrategy.equals(HoodieRecordMerger.PAYLOAD_BASED_MERGE_STRATEGY_UUID)) {
-      return Option.of(HoodieAvroRecordMerger.INSTANCE);
+                                                                   String mergerImpls,
+                                                                   String recordMergerStrategy,
+                                                                   Option<String> payloadClassOpt) {
+    String payloadClass = payloadClassOpt.get();
+    // Payload based strategy and payload class is given.
+    if (null != recordMergerStrategy
```

Comment:
payload class is a table config, shouldn't we fix `HoodieTableConfig.inferRecordMergeModeFromPayloadClass` and `HoodieTableConfig.inferRecordMergeModeFromMergeStrategyId` instead of here?